// @nestjsx/crud-request/lib/request-query.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("@nestjsx/util");
const request_query_validator_1 = require("./request-query.validator");
class RequestQueryBuilder {
    constructor() {
        this._fields = [];
        this._filter = [];
        this._or = [];
        this._join = [];
        this._sort = [];
    }
    static setOptions(options) {
        RequestQueryBuilder._options = Object.assign({}, RequestQueryBuilder._options, options, { paramNamesMap: Object.assign({}, RequestQueryBuilder._options.paramNamesMap, (options.paramNamesMap ? options.paramNamesMap : {})) });
    }
    static getOptions() {
        return RequestQueryBuilder._options;
    }
    static create(params) {
        const qb = new RequestQueryBuilder();
        return util_1.isObject(params) ? qb.createFromParams(params) : qb;
    }
    get options() {
        return RequestQueryBuilder._options;
    }
    query() {
        this.queryString = (this.getJoin() +
            this.getCondition('filter') +
            this.getCondition('or') +
            this.getFields() +
            this.getSort() +
            this.getNumeric('limit') +
            this.getNumeric('offset') +
            this.getNumeric('page') +
            this.getNumeric('cache')).slice(0, -1);
        return this.queryString;
    }
    select(fields) {
        request_query_validator_1.validateFields(fields);
        this._fields = fields;
        return this;
    }
    setFilter(filter) {
        request_query_validator_1.validateCondition(filter, 'filter');
        this._filter.push(filter);
        return this;
    }
    setOr(or) {
        request_query_validator_1.validateCondition(or, 'or');
        this._or.push(or);
        return this;
    }
    setJoin(join) {
        request_query_validator_1.validateJoin(join);
        this._join.push(join);
        return this;
    }
    sortBy(sort) {
        request_query_validator_1.validateSort(sort);
        this._sort.push(sort);
        return this;
    }
    setLimit(limit) {
        request_query_validator_1.validateNumeric(limit, 'limit');
        this._limit = limit;
        return this;
    }
    setOffset(offset) {
        request_query_validator_1.validateNumeric(offset, 'offset');
        this._offset = offset;
        return this;
    }
    setPage(page) {
        request_query_validator_1.validateNumeric(page, 'page');
        this._page = page;
        return this;
    }
    resetCache() {
        this._cache = 0;
        return this;
    }
    createFromParams(params) {
        if (util_1.isArrayFull(params.fields)) {
            this.select(params.fields);
        }
        if (util_1.isArrayFull(params.filter)) {
            params.filter.forEach((filter) => this.setFilter(filter));
        }
        if (util_1.isArrayFull(params.or)) {
            params.or.forEach((or) => this.setOr(or));
        }
        if (util_1.isArrayFull(params.join)) {
            params.join.forEach((join) => this.setJoin(join));
        }
        if (util_1.isArrayFull(params.sort)) {
            params.sort.forEach((sort) => this.sortBy(sort));
        }
        if (!util_1.isNil(params.limit)) {
            this.setLimit(params.limit);
        }
        if (!util_1.isNil(params.offset)) {
            this.setOffset(params.offset);
        }
        if (!util_1.isNil(params.page)) {
            this.setPage(params.page);
        }
        if (params.resetCache) {
            this.resetCache();
        }
        return this;
    }
    getParamName(param) {
        const name = this.options.paramNamesMap[param];
        return util_1.isString(name) ? name : name[0];
    }
    getFields() {
        if (!util_1.hasLength(this._fields)) {
            return '';
        }
        const param = this.getParamName('fields');
        const value = this._fields.join(this.options.delimStr);
        return `${param}=${value}&`;
    }
    getCondition(cond) {
        if (!util_1.hasLength(this[`_${cond}`])) {
            return '';
        }
        const param = this.getParamName(cond);
        const d = this.options.delim;
        const br = this.addBrackets(this[`_${cond}`]);
        return (this[`_${cond}`]
            .map((f) => `${param}${br}=${f.field}${d}${f.operator}${util_1.hasValue(f.value) ? d + f.value : ''}`)
            .join('&') + '&');
    }
    getJoin() {
        if (!util_1.hasLength(this._join)) {
            return '';
        }
        const param = this.getParamName('join');
        const d = this.options.delim;
        const ds = this.options.delimStr;
        const br = this.addBrackets(this._join);
        return (this._join
            .map((j) => `${param}${br}=${j.field}${util_1.isArrayFull(j.select) ? d + j.select.join(ds) : ''}`)
            .join('&') + '&');
    }
    getSort() {
        if (!util_1.hasLength(this._sort)) {
            return '';
        }
        const param = this.getParamName('sort');
        const ds = this.options.delimStr;
        const br = this.addBrackets(this._sort);
        return (this._sort
            .map((s) => `${param}${br}=${s.field}${ds}${s.order}`)
            .join('&') + '&');
    }
    getNumeric(num) {
        if (util_1.isNil(this[`_${num}`])) {
            return '';
        }
        const param = this.getParamName(num);
        const value = this[`_${num}`];
        return `${param}=${value}&`;
    }
    addBrackets(arr) {
        return arr.length > 1 ? '[]' : '';
    }
}
RequestQueryBuilder._options = {
    delim: '||',
    delimStr: ',',
    paramNamesMap: {
        fields: ['fields', 'select'],
        filter: 'filter',
        or: 'or',
        join: 'join',
        sort: 'sort',
        limit: ['per_page', 'limit'],
        offset: 'offset',
        page: 'page',
        cache: 'cache',
    },
};
exports.RequestQueryBuilder = RequestQueryBuilder;
